File: MetaGeta/MetaGeta.GUI/Editors/DirectoryListEditor.vb
Imports System.Windows.Forms
Imports System.Windows.Interop

Namespace Editors
    Public Class DirectoryListEditor
        Inherits ItemListEditor

        Protected Overrides Function CreateItem() As String
            Dim fbd As New FolderBrowserDialog()
            fbd.ShowNewFolderButton = True
            Dim owner As New Win32Window(CType(PresentationSource.FromVisual(Me), HwndSource).Handle)
            If fbd.ShowDialog(owner) <> DialogResult.OK Then Return Nothing
            Return fbd.SelectedPath
        End Function

        Private Class Win32Window
            Implements System.Windows.Forms.IWin32Window

            Private ReadOnly m_Handle As IntPtr

            Public Sub New(ByVal handle As IntPtr)
                m_Handle = handle
            End Sub

            Public ReadOnly Property Handle() As System.IntPtr Implements System.Windows.Forms.IWin32Window.Handle
                Get
                    Return m_Handle
                End Get
            End Property
        End Class

    End Class
End Namespace

File: MetaGeta/MetaGeta.GUI/Editors/DirectoryListEditor.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Windows;
using System.Windows.Forms;
using System.Windows.Interop;
using IWin32Window=System.Windows.Forms.IWin32Window;

#endregion

namespace MetaGeta.GUI.Editors {
    public class DirectoryListEditor : ItemListEditor {
        protected override string CreateItem() {
            var fbd = new FolderBrowserDialog();
            fbd.ShowNewFolderButton = true;
            var owner = new Win32Window(((HwndSource) PresentationSource.FromVisual(this)).Handle);
            if (fbd.ShowDialog(owner) != DialogResult.OK)
                return null;
            return fbd.SelectedPath;
        }

        #region Nested type: Win32Window

        private class Win32Window : IWin32Window {
            private readonly IntPtr m_Handle;

            public Win32Window(IntPtr handle) {
                m_Handle = handle;
            }

            #region IWin32Window Members

            public IntPtr Handle {
                get { return m_Handle; }
            }

            #endregion
        }

        #endregion
    }
}